test/watch.js

```diff
@@ -0,0 +1,318 @@
+const test = require('brittle')
+const { create, createRange, createStored, eventFlush, sleep } = require('./helpers')
+
+test('basic watch', async function (t) {
+  t.plan(2)
+
+  const db = create()
+
+  const watcher = db.watch()
+  t.teardown(() => watcher.destroy())
+
+  watcher.on('change', function (newVersion, oldVersion) {
+    t.is(newVersion, 2)
+    t.is(oldVersion, 1)
+  })
+
+  await db.put('/a')
+})
+
+test('basic watch with onchange option on first arg', async function (t) {
+  t.plan(1)
+
+  const db = create()
+  const onchange = () => t.pass('change')
+
+  const watcher = db.watch(onchange)
+  t.teardown(() => watcher.destroy())
+
+  await db.put('/a')
+})
+
+test('basic watch with onchange option on second arg', async function (t) {
+  t.plan(1)
+
+  const db = create()
+  const onchange = () => t.pass('change')
+
+  const watcher = db.watch({}, onchange)
+  t.teardown(() => watcher.destroy())
+
+  await db.put('/a')
+})
+
+test('basic watch on range', async function (t) {
+  t.plan(1)
+
+  const db = await createRange(50)
+
+  const watcher = db.watch({ gte: '14' })
+  t.teardown(() => watcher.destroy())
+
+  const onchangefail = () => t.fail('should not trigger changes')
+  const onchangepass = () => t.pass('change')
+
+  watcher.on('change', onchangefail)
+  await db.put('13')
+  await eventFlush()
+  await sleep(500)
+  watcher.off('change', onchangefail)
+
+  watcher.on('change', onchangepass)
+  await db.put('14')
+  await eventFlush()
+  watcher.off('change', onchangepass)
+})
+
+test('batch multiple changes', async function (t) {
+  t.plan(2)
+
+  const db = create()
+
+  const watcher = db.watch()
+  t.teardown(() => watcher.destroy())
+
+  watcher.on('change', function (newVersion, oldVersion) {
+    t.is(newVersion, 4)
+    t.is(oldVersion, 1)
+  })
+
+  const batch = db.batch()
+  await batch.put('/a')
+  await batch.put('/b')
+  await batch.put('/c')
+  await batch.flush()
+})
+
+test('watch ready step should not trigger changes if already had entries', async function (t) {
+  t.plan(3)
+
+  const create = createStored()
+
+  const bee = create()
+  await bee.put('/a')
+  await bee.put('/b')
+  await bee.close()
+
+  const db = create()
+  t.is(db.version, 1)
+
+  db.watch(function () {
+    t.fail('should not trigger changes')
+  })
+
+  await db.ready()
+  t.is(db.version, 3)
+
+  await eventFlush()
+  await sleep(500)
+
+  await db.close()
+
+  t.pass()
+})
+
+test('watch without bee.ready() should trigger the correct version changes', async function (t) {
+  t.plan(4)
+
+  const create = createStored()
+
+  const bee = create()
+  await bee.put('/a')
+  await bee.put('/b')
+  await bee.close()
+
+  const db = create()
+  t.is(db.version, 1)
+
+  db.watch(function (newVersion, oldVersion) {
+    t.is(newVersion, 4)
+    t.is(oldVersion, 3)
+  })
+
+  await db.put('/c')
+  await eventFlush()
```

Comment:
why is it event flushing here? needs a comment

Reply:
same for the other ones above

Reply:
That could be an issue as watcher is not graceful when it's destroyed

First, when doing a `db.put` it needs an event flush for watch to trigger a `change` event

With that in mind, when you do `await db.close()` all watchers gets instantly destroyed, so without waiting an event flush there will be no `change` event

Reply:
So the question is, should we support triggering a change event on this case:
```js
db.watch(function (newVersion, oldVersion) {
  // change?
})

await db.put('/c')

await db.close()
```

Sounds like it should

Reply:
Hm even if it looks like it should, adding a simple event flush on the close would not be enough for a diff stream with networking involved. We could make watcher destroy async but making `bee.close` to wait doesn't sound good

Reply:
Maybe we can special case it only for the writter?